infra/src/main/kotlin/fi/fta/geoviite/infra/publication/PublicationDao.kt

```diff
@@ -454,6 +454,98 @@ class PublicationDao(
         }.toMap().also { logger.daoAccess(FETCH, TrackNumberChanges::class, publicationId) }
     }
 
+    fun fetchPublicationLocationTrackSwitchLinkChanges(
+        publicationId: IntId<Publication>,
+    ): Map<IntId<LocationTrack>, LocationTrackPublicationSwitchLinkChanges> =
+        fetchPublicationLocationTrackSwitchLinkChanges(publicationId, null, null)[publicationId] ?: mapOf()
+
+    fun fetchPublicationLocationTrackSwitchLinkChanges(
+        publicationId: IntId<Publication>?,
+        from: Instant?,
+        to: Instant?,
+    ): Map<IntId<Publication>, Map<IntId<LocationTrack>, LocationTrackPublicationSwitchLinkChanges>> {
+        val sql = """
+            select
+              change_side,
+              plt.publication_id,
+              location_track_id,
+              switch_version.id as switch_id,
+              switch_version.name as switch_name,
+              switch_version.external_id as switch_oid
+              from publication.publication
+                join publication.location_track plt on publication.id = plt.publication_id
+                join lateral
+                (select 'new' as change_side, topology_start_switch_id, topology_end_switch_id, alignment_id, alignment_version
+                   from layout.location_track_version ltv
+                   where plt.location_track_id = ltv.id and plt.location_track_version = ltv.version
+                 union all
+                 select 'old', topology_start_switch_id, topology_end_switch_id, alignment_id, alignment_version
+                   from layout.location_track_version ltv
+                   where plt.location_track_id = ltv.id
+                     and plt.location_track_version = ltv.version + 1
+                     and not ltv.draft) ltvs on (true)
```

Comment:
Tämä join-logiikka kaipaisi mielestäni vähän selittämistä. Miksi `old`-puoli on se, jossa on joinissa tuoreempi versio (`ltv.version + 1`?) Lisäksi tuo `+ 1` on mielestäni vähän vaarallinen, sillä vaikka täällä vissiin pelataankin pelkästään officialeilla, en ole varma onko aina takeita siitä että version + 1 palauttaisi aina halutun version. Toimiiko tämä esimerkiksi oikein jos `publicationId`:ksi annetaan uusimman julkaisun id? Lisäksi tämä tämä hakulogiikka ei taitaisi toimia mikäli sitä koskaan haluttaisiin soveltaa `DRAFT`-maailmaan (en tiedä onko sille koskaan tarvetta, mutta sinällään voisin itse nähdä arvoa esim. esikatselunäkymässä virallisen version ja uusimman draftin välin tarkastelussa.) Suosin itse noista syistä aikaleimapohjaisia hakuja aikanaan näitä alunperin kirjoittaessani.

Reply:
Aikalaimapohjaisuus voisi olla kai teoriassa tehtävissä, mutta en katsonut sitä tähän yhteyteen tarpeelliseksi, kun ollaan katsomassa vain julkaisulokin asioita. Tai jos joku keksii sellaisen tilanteen, jossa on käytännön mahdollisuus, että  publicationId:t ei mene julkaisuaikajärjestyksessä, niin sitä vartenhan pitäisi käytä julkaisuaikaleimoja hakemassa.

Versioalgebran selkeydestä en oikein tiedä, mutta selvästi siinä on selkeysongelmia. Ehkä voisi käyttää taulujen koko nimiä.

Reply:
Periaatteessa kantageneroidun ID:n järjestykseen ei pitäisi yleensä luottaa, koska kannat lupaa niille vain uniikkiuden, ei järjestystä. ID:t voidaan varata sekvenssistä epäintuitiivisesti ainakin jos on kyeessä replikoitu kanta (ei ehkä postgresilla ongelma koska vain yksi kirjoittava noodi?) tai jos transaktioita ajetaan rinnakain (ei varmaan julkaisuilla ongelma koska niissä on kantalukko rajaamassa). Jos puhutaan julkaisun järjestyksestä en nyt keksi suoranaista syytä miksei tuo tässä toimisi mutta yleisesti aikaleima lienee ID:tä puhdasoppisempi kronologinen järjestys. https://dba.stackexchange.com/questions/266405/does-ordering-by-auto-incrementing-pk-ensure-chronological-order

Sen sijaan tuo versioiden +1 lienee ok: jos itemistä tehdään uusi draft, sillä on oma id:nsä ja siten oma versiointinsa. Ainoa tapaus jossa draftien ja officialeiden versioita on samalla id:llä on se kun draft on luotu uutena ja tämä blokkaantuu tuolla `not ltv.draft` ehdolla (jos versio -1 on draft, ei voi olla olemassa vanhempaa officialia). En keksi syytä etteikö tuo toimisi. Sinällään versiolla joinissa on aika iso etu luultavasti nopeudessa, sillä se on suora pääavain join versiotauluun.